shared/tinyusb/mp_usbd_cdc.c

```diff
@@ -125,6 +137,13 @@ mp_usbd_line_state_cb
 tud_cdc_line_state_cb
 #endif
     (uint8_t itf, bool dtr, bool rts) {
+    if (dtr) {
+        // The host has started to open the cdc serial port.
+        // Wait a few ms for host to be ready then send tx buffer.
+        cdc_connected_flush_delay = (tud_speed_get() == TUSB_SPEED_HIGH) ? 64 : 8;
+        tud_sof_cb_enable(true);
```

Comment:
Kind of surprised you need to do this, the write_flush() call only prepares the IN endpoint for the host to request it and I'd have thought that the host driver would be prepared to cleanly receive bytes sent after this point. What behaviour do you see if you flush from here, the data goes missing?

EDIT: Oh, I see stm32/usbd_cdc_interface.c does exactly the same thing and has a comment `The actual connection state is delayed to give the host a chance to configure its serial port (in most cases to disable local echo)`. I see in the case of stm32 it delays marking the port as connected at all, so maybe this is for some other purpose that doesn't apply here? As I don't see how local echo matters for data going into the host...?

Reply:
Oh right, maybe it's actually this:
https://github.com/pyserial/pyserial/blob/master/serial/serialposix.py#L351

pyserial default behaviour is to discord the contents of the input buffer immediately on open. :bomb:

Reply:
Yes you've found the pieces of the puzzle - I also found that other non-python applications also often have a similar flush/discard at the end of the connection process, as might be seen in the ramblings on https://github.com/hathach/tinyusb/discussions/2595

So yeah it's simply a delay to give the PC application time to finish doing it's port open and clear the buffers before we send our carefully stored data.